gunicorn keeps booting workers when exceptions raised at startup events

Hi, I have the same issues  when deploy FastAPI using gunicorn + uvicorn，it keeps booting workers when exceptions raised at startup events.

**main.py**

```python
from fastapi import FastAPI

app = FastAPI()


@app.on_event("startup")
async def startup():
    raise Exception("error")


@app.get("/")
async def hello():
    return {"msg": "hello world"}
```

**gunicorn.conf.py**

```python
import json
import os

worker_class = "uvicorn.workers.UvicornWorker"

host = os.getenv("HOST", "0.0.0.0")
port = os.getenv("PORT", "8000")
bind_env = os.getenv("BIND", None)
use_loglevel = os.getenv("LOG_LEVEL", "info")
if bind_env:
    use_bind = bind_env
else:
    use_bind = f"{host}:{port}"


accesslog_var = os.getenv("ACCESS_LOG", "-")
use_accesslog = accesslog_var or None
errorlog_var = os.getenv("ERROR_LOG", "-")
use_errorlog = errorlog_var or None
graceful_timeout_str = os.getenv("GRACEFUL_TIMEOUT", "120")
timeout_str = os.getenv("TIMEOUT", "120")
keepalive_str = os.getenv("KEEP_ALIVE", "5")

# Gunicorn config variables
loglevel = use_loglevel
bind = use_bind
errorlog = use_errorlog
accesslog = use_accesslog
graceful_timeout = int(graceful_timeout_str)
timeout = int(timeout_str)
keepalive = int(keepalive_str)

# For debugging and testing
log_data = {
    "loglevel": loglevel,
    "bind": bind,
    "graceful_timeout": graceful_timeout,
    "timeout": timeout,
    "keepalive": keepalive,
    "errorlog": errorlog,
    "accesslog": accesslog,
}
print(json.dumps(log_data))
```

**gunicorn log**

```text
{"loglevel": "info", "bind": "0.0.0.0:8000", "graceful_timeout": 120, "timeout": 120, "keepalive": 5, "errorlog": "-", "accesslog": "-", "host": "0.0.0.0", "port": "8000"}
[2021-06-01 21:13:05 +0800] [40946] [INFO] Starting gunicorn 20.0.4
[2021-06-01 21:13:05 +0800] [40946] [INFO] Listening at: http://0.0.0.0:8000 (40946)
[2021-06-01 21:13:05 +0800] [40946] [INFO] Using worker: uvicorn.workers.UvicornWorker
[2021-06-01 21:13:05 +0800] [40949] [INFO] Booting worker with pid: 40949
[2021-06-01 21:13:05 +0800] [40949] [INFO] Started server process [40949]
[2021-06-01 21:13:05 +0800] [40949] [INFO] Waiting for application startup.
[2021-06-01 21:13:05 +0800] [40949] [ERROR] Traceback (most recent call last):
  File "./envs/test/lib/python3.7/site-packages/starlette/routing.py", line 526, in lifespan
    async for item in self.lifespan_context(app):
  File "./envs/test/lib/python3.7/site-packages/starlette/routing.py", line 467, in default_lifespan
    await self.startup()
  File "./envs/test/lib/python3.7/site-packages/starlette/routing.py", line 502, in startup
    await handler()
  File "./app/main.py", line 16, in startup
    raise Exception("error")
Exception: error

[2021-06-01 21:13:05 +0800] [40949] [ERROR] Application startup failed. Exiting.
[2021-06-01 21:13:05 +0800] [40949] [INFO] Worker exiting (pid: 40949)
[2021-06-01 21:13:05 +0800] [40950] [INFO] Booting worker with pid: 40950
[2021-06-01 21:13:06 +0800] [40950] [INFO] Started server process [40950]
[2021-06-01 21:13:06 +0800] [40950] [INFO] Waiting for application startup.
[2021-06-01 21:13:06 +0800] [40950] [ERROR] Traceback (most recent call last):
  File "./envs/test/lib/python3.7/site-packages/starlette/routing.py", line 526, in lifespan
    async for item in self.lifespan_context(app):
  File "./envs/test/lib/python3.7/site-packages/starlette/routing.py", line 467, in default_lifespan
    await self.startup()
  File "./envs/test/lib/python3.7/site-packages/starlette/routing.py", line 502, in startup
    await handler()
  File "./app/main.py", line 16, in startup
    raise Exception("error")
Exception: error

[2021-06-01 21:13:06 +0800] [40950] [ERROR] Application startup failed. Exiting.
```

I hope the gunicorn would exit when exceptions raised, but it keeps booting workers. Can someone helps me, thanks?

_Originally posted by @fanchunke1991 in https://github.com/encode/uvicorn/issues/1031#issuecomment-852121760_

--- a/uvicorn/workers.py
+++ b/uvicorn/workers.py
@@ -4,6 +4,7 @@ import signal
 import sys
 from typing import Any
 
+from gunicorn.arbiter import Arbiter
 from gunicorn.workers.base import Worker
 
 from uvicorn.config import Config
@@ -80,7 +81,7 @@ class UvicornWorker(Worker):
         # can shut it down to avoid infinite start/stop cycles.
         # See: https://github.com/encode/uvicorn/issues/1066
         if not server.started:
-            sys.exit(3)
+            sys.exit(Arbiter.WORKER_BOOT_ERROR)
 
     async def callback_notify(self) -> None:
         self.notify()

--- a/uvicorn/workers.py
+++ b/uvicorn/workers.py
@@ -76,7 +76,7 @@ class UvicornWorker(Worker):
         server = Server(config=self.config)
         loop = asyncio.get_event_loop()
         loop.run_until_complete(server.serve(sockets=self.sockets))
-        # Exit with status 3 when worker starts failed, so Gunicorn 
+        # Exit with status 3 when worker starts failed, so Gunicorn
         # can shut it down to avoid infinite start/stop cycles.
         # See: https://github.com/encode/uvicorn/issues/1066
         if not server.started:

--- a/uvicorn/workers.py
+++ b/uvicorn/workers.py
@@ -77,7 +77,7 @@ class UvicornWorker(Worker):
         loop = asyncio.get_event_loop()
         loop.run_until_complete(server.serve(sockets=self.sockets))
         # Exit with status 3 when worker starts failed, so Gunicorn 
-        # can shut it down to aviod infinite start/stop cycles.
+        # can shut it down to avoid infinite start/stop cycles.
         # See: https://github.com/encode/uvicorn/issues/1066
         if not server.started:
             sys.exit(3)

--- a/uvicorn/workers.py
+++ b/uvicorn/workers.py
@@ -1,6 +1,7 @@
 import asyncio
 import logging
 import signal
+import sys
 from typing import Any
 
 from gunicorn.workers.base import Worker
@@ -75,6 +76,11 @@ class UvicornWorker(Worker):
         server = Server(config=self.config)
         loop = asyncio.get_event_loop()
         loop.run_until_complete(server.serve(sockets=self.sockets))
+        # Exit with status 3 when worker starts failed, so Gunicorn 
+        # can shut it down to aviod infinite start/stop cycles.
+        # See: https://github.com/encode/uvicorn/issues/1066
+        if not server.started:
+            sys.exit(3)
 
     async def callback_notify(self) -> None:
         self.notify()

